File: px-checkout/src/main/java/com/mercadopago/android/px/internal/base/BasePresenter.java
package com.mercadopago.android.px.internal.base;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import com.mercadopago.android.px.tracking.internal.TrackWrapper;
import com.mercadopago.android.px.tracking.internal.TrackingContract;
import com.mercadopago.android.px.tracking.internal.events.AbortEvent;
import com.mercadopago.android.px.tracking.internal.events.BackEvent;
import java.lang.ref.WeakReference;

/**
 * Base class for all <code>BasePresenter</code> implementations.
 */

@SuppressWarnings("AbstractClassWithoutAbstractMethods")
public abstract class BasePresenter<V extends MvpView> {

    @Nullable private WeakReference<V> mView;
    @Nullable protected TrackWrapper viewTracker;

    protected final TrackingContract tracker = new TrackingContract() {
        @Override
        public void trackAbort() {
            if (viewTracker != null) {
                new AbortEvent(viewTracker).track();
            }
        }

        @Override
        public void trackBack() {
            if (viewTracker != null) {
                new BackEvent(viewTracker).track();
            }
        }
    };

    protected final void setCurrentViewTracker(@NonNull final TrackWrapper viewTracker) {
        this.viewTracker = viewTracker;
        viewTracker.track();
    }

    public void recoverFromBundle(@NonNull final Bundle bundle) {
    }

    @NonNull
    public Bundle storeInBundle(@NonNull final Bundle bundle) {
        return bundle;
    }

    public void attachView(final V view) {
        mView = new WeakReference<>(view);
    }

    public void detachView() {
        if (mView != null) {
            mView.clear();
            mView = null;
        }
    }

    public boolean isViewAttached() {
        return mView != null && mView.get() != null;
    }

    @NonNull
    public V getView() {
        if (!isViewAttached()) {
            throw new IllegalStateException("view not attached");
        } else {
            return mView.get();
        }
    }
}

File: px-checkout/src/main/java/com/mercadopago/android/px/internal/base/BasePresenter.kt
package com.mercadopago.android.px.internal.base

import com.mercadopago.android.px.tracking.internal.MPTracker
import com.mercadopago.android.px.tracking.internal.TrackWrapper
import com.mercadopago.android.px.tracking.internal.events.AbortEvent
import com.mercadopago.android.px.tracking.internal.events.BackEvent
import java.lang.ref.WeakReference

/**
 * Base class for all `BasePresenter` implementations.
 */
internal abstract class BasePresenter<V : MvpView>(protected val tracker: MPTracker) {

    private var viewReference: WeakReference<V>? = null
    protected var viewTrack: TrackWrapper? = null

    protected val isViewAttached: Boolean
        get() = viewReference?.get() != null
    protected val view: V
        get() = if (!isViewAttached) {
            throw IllegalStateException("view not attached")
        } else {
            viewReference!!.get()!!
        }

    protected fun trackAbort() {
        viewTrack?.let {
            tracker.track(AbortEvent(it))
        }
    }

    protected fun trackBack() {
        viewTrack?.let {
            tracker.track(BackEvent(it))
        }
    }

    protected fun track(trackWrapper: TrackWrapper) {
        tracker.track(trackWrapper)
    }

    protected fun setCurrentViewTracker(viewTrack: TrackWrapper) {
        this.viewTrack = viewTrack
        track(viewTrack)
    }

    open fun attachView(view: V) {
        viewReference = WeakReference(view)
    }

    fun detachView() {
        viewReference?.clear()
        viewReference = null
    }
}
